daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1104,10 +1106,21 @@ private[lf] object SBuiltin {
       val (actualTemplateId, record) = getSAnyContract(args, 1)
 
       if ((templateId +: acceptedTemplateIds).contains(actualTemplateId)) {
-        // TODO: https://github.com/digital-asset/daml/issues/16151
-        // Later, this will need to extend values of predecessor template
-        // types (e.g. by adding the right number of 'None's for missing 'Optional' fields)
-        Control.Value(record.copy(id = templateId))
+        // Here we extend values of predecessor template types by adding the
+        // right number of 'None's for missing 'Optional' fields
```

Comment:
So we're just adding to the end? Are we forcing that new fields must be last, rather than marked as explicitly "new"?
I could see a situation where, say we have
```
template Debt with
    amount : Int
    owedTo : Party
```
might get extended like
```
template Debt with
    amount : Int
    currency : Currency
    owedTo : Party
```
It might be a bit of an odd UX if we also force new fields to be last, but perhaps thats okay

Reply:
we do require new fields to be last, at least for now

Reply:
I think its fine if we have a plan to change it, but I don't think we should allow this restriction to reach users.

Reply:
after discussing with @nickchapman-da, I'll drop this restriction

Reply:
@samuel-williams-da wrote
> It might be a bit of an odd UX if we also force new fields to be last

We should think of the UX of the full daml application as a whole. 
Does reordering fields for clients that do not have access to the Daml type inference will provide a good UX ?